### DISCUSSION TITLE
Is calling `ref.watch(myAuthProvider.future)` inside of a `GoRouter.redirect` a bad practice?

### DISCUSSION THREAD

**Discussion by sashkent3** [rrousselGit/riverpod#3751] [Q&A]
I've been in search of a **simple** solution for integrating `riverpod` with the `go_router` package in terms of reacting to authentication state changes. I've since seen a lot of conflicting complicated solutions none of which were satisfactory.
It [seems](https://github.com/flutter/packages/blob/e9253e60161b498f2528a76198b6cd57a1a0d252/packages/go_router/lib/src/router.dart#L63) that from the `go_router` side the recommended solution is to utilize `BuildContext.dependOnInheritedWidgetOfExactType` inside of the `GoRouter.redirect` function. My question then is then twofold:
1. Does the `ref.watch()` call `BuildContext.dependOnInheritedWidgetOfExactType` internally?
2. What happens if I call `ref.watch()` inside of a `GoRouter.redirect`?

**Comment by rrousselGit** [comment]
> Does the ref.watch() call BuildContext.dependOnInheritedWidgetOfExactType internally?

Yes, but actually no :shrug:
It _does_ call that method. But that's not going to do what you want.

The part that triggers a rebuild of a Consumer isn't done through `BuildContext.dependOnInheritedWidgetOfExactType`, but instead calling `Element.markNeedsBuild`.
As such, go_router natively has no idea that a provider changed for now.


> What happens if I call ref.watch() inside of a GoRouter.redirect?

GoRouter doesn't give you a ref. A Consumer did. What will happen is, the Consumer that gave you the Ref will rebuild.

**Reply by sashkent3** [reply]
If I understood you correctly, the `ref.watch()` approach will always cause the `GoRouter` to be rebuilt on provider state changes which is undesirable. It then seems to me that it's best/simplest to use a `provider`'s `Provider` for the authentication state. I can then use `riverpod`'s `ChangeNotifierProvider` adapter for all of my `riverpod` needs. Any downsides?

**Reply by rrousselGit** [reply]
ref.watch doesn't cause GoRouter to rebuild. It causes your Consumer to rebuild. GoRouter naturally has no clue that something changed

**Reply by sashkent3** [reply]
I mean the app would look something like:
```
class MyApp extends ConsumerWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp.router(
      routerConfig: GoRouter(
        redirect: (context, state) async {
          final user = ref.watch(myAuthProvider.provider);
        },
      )
    );
  }
}
```

Here every time the auth state changes we rebuild the whole app including the `GoRouter`, correct?
